drivers/power_domain/power_domain_state_change_device_onoff.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/kernel_structs.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
+#include <zephyr/pm/pm.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(power_domain_deviceonoff, CONFIG_POWER_DOMAIN_LOG_LEVEL);
+
+/* Indicates the end of the onoff_power_states array for count purposes */
+#define POWER_DOMAIN_DEVICE_ONOFF_STATE_COUNT 0xFF
+
+struct pd_deviceonoff_config {
+	uint8_t *onoff_power_states;
+};
+
+struct pd_visitor_context {
+	const struct device *domain;
+	enum pm_device_action action;
+};
+
+static int pd_domain_visitor(const struct device *dev, void *context)
+{
+	struct pd_visitor_context *visitor_context = context;
+
+	/* Only run action if the device is on the specified domain */
+	if (!dev->pm || (dev->pm_base->domain != visitor_context->domain)) {
+		return 0;
+	}
+
+	/* In case device is active, first suspend it before turning it off */
+	if ((visitor_context->action == PM_DEVICE_ACTION_TURN_OFF) &&
+	    (dev->pm_base->state == PM_DEVICE_STATE_ACTIVE)) {
+		(void)pm_device_action_run(dev, PM_DEVICE_ACTION_SUSPEND);
+	}
+	(void)pm_device_action_run(dev, visitor_context->action);
+	return 0;
+}
+
+static int pd_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	const struct pd_deviceonoff_config *config = dev->config;
+	uint8_t i = 0;
+	/* Get the next power state that will be used */
+	enum pm_state state = pm_state_next_get(_current_cpu->id)->state;
+	struct pd_visitor_context context = {.domain = dev};
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		while (config->onoff_power_states[i] != POWER_DOMAIN_DEVICE_ONOFF_STATE_COUNT) {
+			/* Check if we need do the turn on action for this state */
+			if (state == config->onoff_power_states[i]) {
+				/* Notify devices on the domain they are now powered */
+				context.action = PM_DEVICE_ACTION_TURN_ON;
+				(void)device_supported_foreach(dev, pd_domain_visitor, &context);
+				/* No need to go through the rest of the array of states */
+				break;
+			}
+			i++;
+		}
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
```

Comment:
You've gone to the effort of defining a LOG module, it would be nice to add a `LOG_INF("%s: suspending", dev->name);` or something here and in the `PM_DEVICE_ACTION_RESUME` handling

Reply:
Thanks for the suggestion. I have added a few.